pubsub/pumps.go

```diff
@@ -0,0 +1,127 @@
+package pubsub
+
+import (
+	"net"
+
+	"github.com/garyburd/redigo/redis"
+)
+
+// readPump tries to read out data from the connection, sending it down
+// the data channel, until it's closed. It's a conversion from
+// synchronous to channel-based code.
+type readPump struct {
+	cnx redis.Conn
+
+	data   chan interface{}
+	errs   chan error
+	closer chan struct{}
+}
+
+// newReadPump creates a new pump that operates on the single Redis connection.
+func newReadPump(cnx redis.Conn) *readPump {
+	return &readPump{
+		cnx:    cnx,
+		data:   make(chan interface{}),
+		errs:   make(chan error),
+		closer: make(chan struct{}),
+	}
+}
+
+// Work starts reading from the connection and blocks until it is closed.
+func (r *readPump) Work() {
+	cnx := redis.PubSubConn{Conn: r.cnx}
+	defer close(r.closer)
```

Comment:
Sorry, I missed this comment before. I think leaving them open is preferable in this case; the pump is a structure whose control flow is entirely contained by the Pubsub struct, and not closing channels makes the master select block in L165 of redis.go possible. Otherwise we'd have to use reflect.Select or something along those lines to prune out old channels, or do some other kind of ugliness.